### ISSUE TITLE
Clients on the second screen cover the wibox status bar

### ISSUE THREAD

**Issue by aroig** [awesomeWM/awesome#892]
I just noticed this. Clients on my second screen cover the status bar, while on the first screen they do work just fine (whatever it means first and  second, since screens just come and go now...). Besides that, the tiling is ok, according to the layout.

I can reproduce it with the default configuration (no custom rc.lua). It started happening as a result of a commit in the last week or two.

When I get the time I can try to narrow it down a bit more, but I thought I'd just report it, in case you already know where to look!

**Comment by psychon** [comment]
> (whatever it means first and second, since screens just come and go now...)

1. You already use the "don't restart on RandR changes"-stuff?
2. It actually works? :-)

Does it happen to all clients or just some special one?
Could you run `echo 'local res = "" ; for s=1, screen.count() do res = res .. tostring(s) .. ": " .. require("gears.debug").dump_return({geometry = screen[s].geometry, workarea = screen[s].workarea}) .. "\\n" end return res' | awesome-client` and tell us if the output is what you expect (and actually show us the output)? Geometry is the size of the screen, workarea is the part that clients are layouted on.

If the workarea is actually wrong, does `echo 'client.focus:struts{ top = 4, bottom = 4, left = 4, right = 4}' | awesome-client` fix things? (This kind-of breaks the currently focused client. You should close it afterwards. This is just a cheap hack to force the C code to re-compue workareas)

**Comment by aroig** [comment]
> You already use the "don't restart on RandR changes"-stuff?

Yes, I updated to current master this morning to make sur it still happens. 


> It actually works? :-)

* I tried on my laptop (1 screen) before the weekend and seemed to work. 
* When I tested it today on my desktop (two screens plugged in from the begining), it mostly worked except from what I reported.
* I haven't tested changing screens dynamically yet. 

> Does it happen to all clients or just some special one?

All of them on one one of the screens.

> Could you run echo 'local res = "" ; for s=1, screen.count() do res = res .. tostring(s) .. ": " .. require("gears.debug").dump_return({geometry = screen[s].geometry, workarea = screen[s].workarea}) .. "\\n" end return res' | awesome-client and tell us if the output is what you expect (and actually show us the output)? Geometry is the size of the screen, workarea is the part that clients are layouted on.

    1: table: 0x1389c60
      geometry : table: 0x1389d30
        height : 1080 (number)
        y : 0 (number)
        x : 0 (number)
        width : 1920 (number)
      workarea : table: 0x1e956e0
        height : 1062 (number)
        y : 18 (number)
        x : 0 (number)
        width : 1920 (number)
    2: table: 0x1ae4f20
      geometry : table: 0x1ae4f60
        height : 1080 (number)
        y : 0 (number)
        x : 1920 (number)
        width : 1920 (number)
      workarea : table: 0x1ae4fa0
        height : 1080 (number)
        y : 0 (number)
        x : 1920 (number)
        width : 1920 (number)

The workarea of screen 2 is wrong. it should be `height: 1062` like the first one.


> If the workarea is actually wrong, does echo 'client.focus:struts{ top = 4, bottom = 4, left = 4, right = 4}' | awesome-client fix things? (This kind-of breaks the currently focused client. You should close it afterwards. This is just a cheap hack to force the C code to re-compute workareas)

Yes it does.

Thanks for your time on this!

**Comment by Elv13** [comment]
Hello,

Can you try https://github.com/awesomeWM/awesome/pull/878 ? (I didn't test
yet with multiple screens, so I have no clue if it's going to fix your
issue or make it worst)

Thanks

**Comment by aroig** [comment]
> Can you try #878 ? (I didn't test yet with multiple screens, so I have no clue if it's going to fix your
issue or make it worst)

Ok, so I tested #878 merged on top of master. It sort-of-fixes it. On one hand, the clients now respect the top strut for the wibox on both screens.

On the other hand, there is an additional empty lower strut of the same size at the bottom of both screens. 

    1: table: 0x1cf47e0
      workarea : table: 0x1cf48f0
        y : 18 (number)
        x : 0 (number)
        height : 1044 (number)
        width : 1920 (number)
      geometry : table: 0x1cf4820
        y : 0 (number)
        x : 0 (number)
        height : 1080 (number)
        width : 1920 (number)
    2: table: 0x2336aa0
      workarea : table: 0x2336bb0
        y : 18 (number)
        x : 1920 (number)
        height : 1044 (number)
        width : 1920 (number)
      geometry : table: 0x2336ae0
        y : 0 (number)
        x : 1920 (number)
        height : 1080 (number)
        width : 1920 (number)
 
This time @psychon's black magic (the `client.focus:struts` thing) does not help.

**Comment by Elv13** [comment]
Ah, of course, I didn't check the wibox/wibar screen when computing the margins, so the margins was the sum of the wibars/wiboxes on all screens... I pushed a new commit that fixes it (maybe?).

**Comment by aroig** [comment]
> I pushed a new commit that fixes it (maybe?).

No, It changes nothing for me.

**Comment by psychon** [comment]
No idea from where to where I upgraded to (I'm only using this on weekends; the upgrade must have been last weekend), but I can now reproduce this now, too.

Also, selecting text an a terminal randomly fails and instead the `awful.mouse.snap` Windows-7-style edge snap things it's time to be active (added in commit 91cdc something, I can't copy&paste the commit ID right now due to this exact issue).

**Comment by Elv13** [comment]
The snap is only supposed to trigger when the client is moved using `awful.mouse.client.move`. Can you add some prints to see if (and way) it is being called?

**Comment by psychon** [comment]
Possible fix (yay, this was already broken before (missing `property::workarea` signal when a wibox/drawin was moved to a different screen), we just didn't notice before the "cache workarea"-change):
```patch
diff --git a/objects/drawin.c b/objects/drawin.c
index 8b71106..ed5c08b 100644
--- a/objects/drawin.c
+++ b/objects/drawin.c
@@ -214,6 +214,15 @@ drawin_moveresize(lua_State *L, int udx, area_t geometry)
         luaA_object_emit_signal(L, udx, "property::width", 0);
     if (old_geometry.height != w->geometry.height)
         luaA_object_emit_signal(L, udx, "property::height", 0);
+
+    // https://github.com/awesomeWM/awesome/issues/892
+    screen_t *old_screen = screen_getbycoord(old_geometry.x, old_geometry.y);
+    screen_t *new_screen = screen_getbycoord(w->geometry.x, w->geometry.y);
+    if (old_screen != new_screen && strut_has_value(&w->strut))
+    {
+        screen_update_workarea(old_screen);
+        screen_update_workarea(new_screen);
+    }
 }
 
 /** Refresh the window content by copying its pixmap data to its window.
```
I'm too lazy to restart my awesome instance for actually testing this until tomorrow...
@Elv13 I don't know, but will of course investigate. The terminal selection is a mouse grabber, the mouse cursor changes when I press the mouse button. Also, the Java fix really should not cause anything... meh! :-(

**Comment by aroig** [comment]
I can confirm that current master + @psychon fix above fixes my original issue.

There is still the second strut problem with PR #878. I'll repeat the problem there so it is easier to keep track of.

**Comment by Elv13** [comment]
> Also, selecting text an a terminal randomly fails and instead the awful.mouse.snap Windows-7-style edge snap things it's time to be active (added in commit 91cdc something, I can't copy&paste the commit ID right now due to this exact issue).

Ok, I've seen it. It went away after I merged 394ff06589676c26561f19264de1097edaaecdc7 into my branch.

As for the scroll/select issue in urxvt, it happens when I maximize and unmaximize clients, then click on the terminal. But it doesn't happen all the time. It seems random. It's why I stayed quiet until now. Don't waste too much time on this until I figure out the exact sequence of event that triggers it (I just tried to reproduce it and failed, but it has been happening for a month from time to time).

**Comment by psychon** [comment]
In my case, this was exactly what 394ff06 fixed: The code assumed clicks in the "upper part" of the client were titlebar clicks, which explains why `awful.mouse.client.move` ran. It's all a big input-breakage caused by me (and now already fixed). Sorry.
